packages/@lwc/engine-dom/src/formatters/component.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { getAssociatedVMIfPresent, LightningElement } from '@lwc/engine-core';
+import { isUndefined, ownKeys } from '@lwc/shared';
+
+/**
+ * Displays the header for a custom element.
+ *
+ * @param ce the custom element
+ * @param componentInstance component instance associated with the custom element.
+ */
+function getHeaderForCustomElement(ce: HTMLElement, componentInstance: LightningElement) {
+    // [element]
+    // LWC component instance: [vm.component]
+    return [
+        'div',
+        {},
+        ['object', { object: ce, config: { skip: true } }],
+        [
+            'div',
+            {},
+            ['span', { style: 'margin: 0 5px; color: red' }, 'LWC component instance:'],
+            ['object', { object: componentInstance }],
+        ],
+    ];
+}
+
+function getHeaderForComponentInstance(
+    componentInstance: LightningElement,
+    debugInfo: Record<symbol, any>
+) {
+    if (ownKeys(debugInfo).length === 0) {
+        // there is no debug information, no need to customize this component instance
+        return null;
+    }
+
+    // [component]
+    // Debug information: [vm.debugInfo]
+    return [
+        'div',
+        {},
+        ['object', { object: componentInstance, config: { skip: true } }],
+        [
+            'div',
+            {},
+            ['span', { style: 'margin: 0 5px; color: red' }, 'Debug information:'],
+            ['object', { object: debugInfo }],
+        ],
+    ];
+}
+
+export const LightningElementFormatter = {
+    name: 'LightningElementFormatter',
+
+    header(obj: any, config?: Record<string, any>) {
+        const vm = getAssociatedVMIfPresent(obj);
+
+        if (!isUndefined(vm) && (isUndefined(config) || !config.skip)) {
```

Comment:
what is `config.skip` for?

Reply:
it is used to avoid an infinite loop, because `componentInstance` is displayed as `componentInstance + debugInfo`, and similarly for the custom element.